File: rust/crates/rpg-agents/src/black_swan_agent.rs
//! black_swan_agent — 主动触发世界事件的子代理。
//!
//! 对应 Python: `rpg/agents/black_swan_agent.py`
//!
//! 5 层 validator 管线:
//! 1. reality_snapshot — 现实切片快照(给 LLM)
//! 2. proposal_tool_schema — native tool_use 强 schema
//! 3. validators — 5 个独立校验(token blacklist / hard constraint /
//!    timeline anchor / NPC presence / independent critic)
//! 4. retry — max 2 次
//! 5. dispatcher — 落地,origin="autonomous_agent"
//!
//! ⚠️ MEMORY:rpg_black_swan_agent_todo.md 备忘的就是这个 agent。
//! 设计已就绪,但依赖 task 87 工具化底座。本骨架完成 Layer 1+2+3 接口。

use serde::{Deserialize, Serialize};
use serde_json::{json, Value};

use crate::common::{AgentResult, GameState, SharedLlm, ToolSchema};
use rpg_state::{apply_op, GameState as RealGameState, Op};

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct RealitySnapshot {
    pub current_phase: String,
    pub current_location: String,
    pub current_time: String,
    pub active_npcs: Vec<Value>,
    pub locked_variables: serde_json::Map<String, Value>,
    pub recent_events: Vec<String>,
    pub chapter_window: Value,
    pub turn: u64,
}

/// Layer 1: 现实切片。
pub fn reality_snapshot(state: &GameState, _script_id: Option<i64>) -> RealitySnapshot {
    let world = state.data.get("world").cloned().unwrap_or(Value::Null);
    let timeline = world.get("timeline").cloned().unwrap_or(Value::Null);
    let player = state.data.get("player").cloned().unwrap_or(Value::Null);
    let worldline = state.data.get("worldline").cloned().unwrap_or(Value::Null);

    let mut locked_vars = serde_json::Map::new();
    if let Some(user_vars) = worldline.get("user_variables").and_then(|v| v.as_object()) {
        for (key, info) in user_vars {
            if info.get("locked").and_then(|v| v.as_bool()).unwrap_or(false) {
                let v = info.get("value").cloned().unwrap_or(Value::String("".into()));
                locked_vars.insert(key.clone(), v);
            }
        }
    }

    let active_entities = state
        .data
        .get("active_entities")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default();

    let active_npcs: Vec<Value> = active_entities
        .into_iter()
        .filter(|e| {
            let kind = e.get("kind").and_then(|v| v.as_str()).unwrap_or("unknown");
            matches!(kind, "npc" | "enemy" | "unknown")
        })
        .take(8)
        .map(|e| {
            json!({
                "id": e.get("id").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                "name": e.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string(),
                "disposition": e.get("disposition").and_then(|v| v.as_str()).unwrap_or("unknown").to_string(),
                "kind": e.get("kind").and_then(|v| v.as_str()).unwrap_or("unknown").to_string(),
            })
        })
        .collect();

    let recent_events: Vec<String> = world
        .get("known_events")
        .and_then(|v| v.as_array())
        .cloned()
        .unwrap_or_default()
        .into_iter()
        .rev()
        .take(5)
        .map(|v| v.as_str().unwrap_or("").to_string())
        .collect::<Vec<_>>()
        .into_iter()
        .rev()
        .collect();

    RealitySnapshot {
        current_phase: timeline
            .get("current_phase")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        current_location: player
            .get("current_location")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        current_time: world.get("time").and_then(|v| v.as_str()).unwrap_or("").to_string(),
        active_npcs,
        locked_variables: locked_vars,
        recent_events,
        chapter_window: json!({
            "min": timeline.get("chapter_min").cloned().unwrap_or(Value::Null),
            "max": timeline.get("chapter_max").cloned().unwrap_or(Value::Null),
        }),
        turn: state.turn,
    }
}

/// Layer 2: 生成 LLM tool_use schema(enum 限定 phase/character/location 取值)。
pub fn proposal_tool_schema(snapshot: &RealitySnapshot) -> ToolSchema {
    let valid_npc_ids: Vec<String> = snapshot
        .active_npcs
        .iter()
        .filter_map(|n| n.get("id").and_then(|v| v.as_str()).map(|s| s.to_string()))
        .filter(|s| !s.is_empty())
        .collect();

    let schema = json!({
        "type": "object",
        "required": ["event_kind", "summary"],
        "properties": {
            "event_kind": {
                "type": "string",
                "enum": ["weather", "npc_arrival", "rumor", "encounter", "object", "no_op"],
            },
            "summary": {"type": "string", "description": "Short narrative summary"},
            "affected_npc_ids": {
                "type": "array",
                "items": {"type": "string", "enum": valid_npc_ids},
            },
            "drift_score": {"type": "number", "minimum": 0.0, "maximum": 1.0},
            "rationale": {"type": "string"},
        }
    });

    ToolSchema {
        name: "propose_black_swan_event".into(),
        description: "Propose a black swan event for the current game phase. \
                      Use ONLY entities, locations, and concepts that appear in the snapshot. \
                      DO NOT invent new NPCs, locations, or cross-phase events. \
                      If no suitable event fits the current situation, return event_kind='no_op'."
            .into(),
        input_schema: schema,
    }
}

// ── Layer 3: validators ──────────────────────────────────────────────

/// 3a: token blacklist — proposal 含禁用词直接拒绝。
pub fn validator_token_blacklist(
    proposal: &Value,
    blacklist: &[&str],
) -> (bool, String) {
    let summary = proposal
        .get("summary")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    for kw in blacklist {
        if summary.contains(kw) {
            return (false, format!("命中 blacklist: {kw}"));
        }
    }
    (true, String::new())
}

/// 3c: hard constraint — proposal 不得违反 locked_variables。
pub fn validator_hard_constraints(
    proposal: &Value,
    snapshot: &RealitySnapshot,
) -> (bool, String) {
    let summary = proposal.get("summary").and_then(|v| v.as_str()).unwrap_or("");
    for (key, val) in &snapshot.locked_variables {
        if let Some(v) = val.as_str() {
            if !v.is_empty()
                && !summary.contains(v)
                && summary.contains(key)
            {
                return (
                    false,
                    format!("locked_variable {key}={v} 与 proposal 内容冲突"),
                );
            }
        }
    }
    (true, String::new())
}

/// 3d: timeline anchor — proposal 不得跨 phase。
pub fn validator_timeline_anchor(
    proposal: &Value,
    snapshot: &RealitySnapshot,
) -> (bool, String) {
    let proposed_phase = proposal
        .get("phase")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    if !proposed_phase.is_empty() && proposed_phase != snapshot.current_phase {
        return (
            false,
            format!(
                "proposal phase '{}' 与 current_phase '{}' 不符",
                proposed_phase, snapshot.current_phase
            ),
        );
    }
    (true, String::new())
}

/// 3b: NPC presence — affected_npc_ids 必须都在 snapshot.active_npcs 里。
pub fn validator_npc_presence(
    proposal: &Value,
    snapshot: &RealitySnapshot,
) -> (bool, String) {
    let active_ids: std::collections::HashSet<&str> = snapshot
        .active_npcs
        .iter()
        .filter_map(|n| n.get("id").and_then(|v| v.as_str()))
        .collect();
    let Some(arr) = proposal.get("affected_npc_ids").and_then(|v| v.as_array()) else {
        return (true, String::new());
    };
    for v in arr {
        let id = v.as_str().unwrap_or("");
        if !id.is_empty() && !active_ids.contains(id) {
            return (false, format!("NPC '{id}' 不在 active_npcs"));
        }
    }
    (true, String::new())
}

/// 3e: independent critic — 接独立 LLM 二次判定(本骨架返回通过)。
///
/// TODO[opus]: 引入第二个便宜 LLM 做 critic + reason 校验。
pub fn validator_independent_critic(
    _proposal: &Value,
    _snapshot: &RealitySnapshot,
) -> (bool, String) {
    (true, String::new())
}

/// 串行跑全部 validator。返回 (all_pass, failures)。
pub fn run_validators(
    proposal: &Value,
    snapshot: &RealitySnapshot,
    blacklist: &[&str],
) -> (bool, Vec<String>) {
    let mut failures: Vec<String> = Vec::new();
    let checks: [(bool, String); 5] = [
        validator_token_blacklist(proposal, blacklist),
        validator_hard_constraints(proposal, snapshot),
        validator_timeline_anchor(proposal, snapshot),
        validator_npc_presence(proposal, snapshot),
        validator_independent_critic(proposal, snapshot),
    ];
    for (pass, msg) in checks {
        if !pass {
            failures.push(msg);
        }
    }
    (failures.is_empty(), failures)
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct BlackSwanInput {
    pub user_id: Option<i64>,
    pub script_id: Option<i64>,
    pub probability: f64,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct BlackSwanOutput {
    pub triggered: bool,
    pub proposal: Option<Value>,
    pub validation_failures: Vec<String>,
}

pub struct BlackSwanAgent {
    #[allow(dead_code)] // 当前管线 LLM 提案路径走 GM 端,这里 dispatcher only
    llm: SharedLlm,
}

impl BlackSwanAgent {
    pub fn new(llm: SharedLlm) -> Self {
        Self { llm }
    }

    /// 主入口:Layer 1 snapshot → Layer 2 schema → LLM propose → Layer 3 validate → Layer 5 dispatch。
    ///
    /// 当前 LLM propose 仍走 GM 端 — 本 agent 提供 snapshot + schema + 校验闸门 + dispatch。
    /// 上层若想跑完整 maybe_trigger,需自己拿 schema 调 LLM 再调 dispatch。
    pub async fn maybe_trigger(
        &self,
        input: BlackSwanInput,
        state: &GameState,
    ) -> AgentResult<BlackSwanOutput> {
        // 概率门:用 thread_rng 而非 secrets — 此处只是 trigger / no-trigger 不涉敏感随机。
        let probability = input.probability.clamp(0.0, 1.0);
        if probability <= 0.0 {
            return Ok(BlackSwanOutput {
                triggered: false,
                proposal: None,
                validation_failures: vec!["probability<=0,本轮不触发".into()],
            });
        }
        let roll: f64 = rand::random();
        if roll > probability {
            return Ok(BlackSwanOutput {
                triggered: false,
                proposal: None,
                validation_failures: vec![format!(
                    "概率未命中 (roll={roll:.3} > p={probability:.3})"
                )],
            });
        }

        let snapshot = reality_snapshot(state, input.script_id);
        let _schema = proposal_tool_schema(&snapshot);

        // 完整 propose+validate 路径需 GM 端协同;此处只做接口校验,返回空。
        // 上层若有 proposal,可直接调 dispatch_swan() apply 到 RealGameState。
        Ok(BlackSwanOutput {
            triggered: false,
            proposal: None,
            validation_failures: vec!["snapshot 已生成,LLM propose 待上层调用 dispatch_swan".into()],
        })
    }
}

/// 把校验通过的 proposal 落地到 RealGameState(rpg-state)。
///
/// 对应 Python `_apply_swan`:把 proposal 转成一组 Op,经由 apply_op 闸门写入 state。
/// `origin` 决定 source 标识,黑天鹅 agent 用 "autonomous_agent"。
///
/// 返回 (applied_ops_count, errors)。
pub fn dispatch_swan(
    state: &mut RealGameState,
    proposal: &Value,
    origin: &str,
) -> (u32, Vec<String>) {
    let mut applied = 0u32;
    let mut errors: Vec<String> = Vec::new();
    let event_kind = proposal
        .get("event_kind")
        .and_then(|v| v.as_str())
        .unwrap_or("");
    if event_kind.is_empty() || event_kind == "no_op" {
        return (0, vec!["event_kind=no_op,跳过".into()]);
    }
    let summary = proposal
        .get("summary")
        .and_then(|v| v.as_str())
        .unwrap_or("")
        .to_string();
    let source = if origin.is_empty() {
        "autonomous_agent".to_string()
    } else {
        format!("autonomous_agent:{origin}")
    };

    // 1) world.known_events append summary
    if !summary.is_empty() {
        let op = Op::Append {
            path: "world.known_events".into(),
            value: Value::String(summary.clone()),
        };
        match apply_op(state, op, &source, true) {
            Ok(_) => applied += 1,
            Err(e) => errors.push(format!("world.known_events append 失败: {e}")),
        }
    }

    // 2) world.last_swan = {kind, summary, ts}
    let last_swan = json!({
        "event_kind": event_kind,
        "summary": summary,
        "rationale": proposal.get("rationale").cloned().unwrap_or(Value::Null),
        "ts": chrono::Utc::now().to_rfc3339(),
    });
    let op = Op::Set {
        path: "world.last_swan".into(),
        value: last_swan,
    };
    match apply_op(state, op, &source, true) {
        Ok(_) => applied += 1,
        Err(e) => errors.push(format!("world.last_swan set 失败: {e}")),
    }

    // 3) kind 特化字段 — weather / npc_arrival / rumor 等
    match event_kind {
        "weather" => {
            if !summary.is_empty() {
                let op = Op::Set {
                    path: "world.weather".into(),
                    value: Value::String(summary.clone()),
                };
                if let Err(e) = apply_op(state, op, &source, true) {
                    errors.push(format!("world.weather set 失败: {e}"));
                } else {
                    applied += 1;
                }
            }
        }
        "npc_arrival" => {
            // 把 affected_npc_ids 加进 world.active_npcs
            if let Some(ids) = proposal
                .get("affected_npc_ids")
                .and_then(|v| v.as_array())
            {
                for id in ids {
                    let id_str = id.as_str().unwrap_or("").to_string();
                    if id_str.is_empty() {
                        continue;
                    }
                    let op = Op::Append {
                        path: "world.active_npcs".into(),
                        value: Value::String(id_str),
                    };
                    if let Err(e) = apply_op(state, op, &source, true) {
                        errors.push(format!("world.active_npcs append 失败: {e}"));
                    } else {
                        applied += 1;
                    }
                }
            }
        }
        "rumor" => {
            let op = Op::Append {
                path: "memory.rumors".into(),
                value: Value::String(summary.clone()),
            };
            if let Err(e) = apply_op(state, op, &source, true) {
                errors.push(format!("memory.rumors append 失败: {e}"));
            } else {
                applied += 1;
            }
        }
        "encounter" | "object" => {
            // 通用:写到 memory.encounters
            let op = Op::Append {
                path: format!("memory.{event_kind}s"),
                value: json!({
                    "summary": summary,
                    "ts": chrono::Utc::now().to_rfc3339(),
                }),
            };
            if let Err(e) = apply_op(state, op, &source, true) {
                errors.push(format!("memory.{event_kind}s append 失败: {e}"));
            } else {
                applied += 1;
            }
        }
        _ => {}
    }

    (applied, errors)
}

File: rpg/agents/black_swan_agent.py
"""agents.black_swan_agent — 主动触发世界事件的子代理。

5 层 validator 管线 (设计 ready,本次 MVP 实现 Layer 1+2+3a+3c+3d+4+5):
1. 现实切片快照
2. native tool_use 强 schema
3. 5 个 validator (本次 3 个: 3a token blacklist / 3c hard constraint / 3d timeline anchor;
   3b NPC presence / 3e independent critic 留接口)
4. retry (max 2 次)
5. dispatcher 落地,origin="autonomous_agent"

触发方式: Path B (post-GM chat handler hook)。
"""
from __future__ import annotations

import re
import secrets
from typing import Any, Callable, Optional, cast

# ─── Layer 1: 现实切片 ────────────────────────────────────────────

def reality_snapshot(state: Any, script_id: int | None = None) -> dict:
    """暴露当前剧情快照给 LLM,不含敏感字段。

    返回字段:
      - current_phase: str
      - current_location: str
      - current_time: str
      - active_npcs: list[dict] (id + name + disposition)
      - locked_variables: dict[str, str]
      - recent_events: list[str] (最近 5 条 known_events)
      - chapter_window: dict (script_id 时含 chapter_min/max)
    """
    data = getattr(state, 'data', {}) or {}
    world = data.get('world', {}) or {}
    timeline = world.get('timeline', {}) or {}
    player = data.get('player', {}) or {}
    worldline = data.get('worldline', {}) or {}

    # locked vars (user 硬约束)
    locked_vars: dict[str, str] = {}
    user_vars = worldline.get('user_variables', {}) or {}
    for key, info in user_vars.items():
        if isinstance(info, dict) and info.get('locked'):
            locked_vars[key] = info.get('value', '')

    # active NPCs (轻量索引)
    active_entities = data.get('active_entities', []) or []
    active_npcs = [
        {
            'id': str(e.get('id', '')),
            'name': str(e.get('name', '')),
            'disposition': str(e.get('disposition', 'unknown')),
            'kind': str(e.get('kind', 'unknown')),
        }
        for e in active_entities
        if e.get('kind') in ('npc', 'enemy', 'unknown') or not e.get('kind')
    ][:8]  # 最多 8 个

    # recent events
    known_events = world.get('known_events', []) or []
    recent_events = [str(e) for e in known_events[-5:]]

    return {
        'current_phase': str(timeline.get('current_phase', '')),
        'current_location': str(player.get('current_location', '')),
        'current_time': str(world.get('time', '')),
        'active_npcs': active_npcs,
        'locked_variables': locked_vars,
        'recent_events': recent_events,
        'chapter_window': {
            'min': timeline.get('chapter_min'),
            'max': timeline.get('chapter_max'),
        },
        'turn': data.get('turn', 0),
    }


# ─── Layer 2: 强 schema tool_use ────────────────────────────────

def proposal_tool_schema(snapshot: dict) -> dict:
    """生成 LLM tool_use schema,enum 限定 phase/character/location 取值。

    返回 Anthropic tool_use 兼容的 dict (name + description + input_schema)。
    """
    # 从 snapshot 抽 enum
    valid_npc_ids = [n['id'] for n in snapshot.get('active_npcs', []) if n.get('id')]

    return {
        "name": "propose_black_swan_event",
        "description": (
            "Propose a black swan event for the current game phase. "
            "Use ONLY entities, locations, and concepts that appear in the snapshot. "
            "DO NOT invent new NPCs, locations, or cross-phase events. "
            "If no suitable event fits the current situation, return event_kind='no_op'."
        ),
        "input_schema": {
            "type": "object",
            "required": ["event_kind", "summary"],
            "properties": {
                "event_kind": {
                    "type": "string",
                    "enum": ["new_event", "npc_action", "environment_change", "no_op"],
                },
                "summary": {
                    "type": "string",
                    "description": "1-2 sentence narrative summary (Chinese OK)",
                    "maxLength": 200,
                },
                "involved_npcs": {
                    "type": "array",
                    "items": {"type": "string", "enum": valid_npc_ids or [""]},
                    "default": [],
                },
                "location": {
                    "type": "string",
                    "description": "Must be current_location or a sub-location of it",
                },
                "tools_to_call": {
                    "type": "array",
                    "description": "Optional: tool calls to dispatch (e.g. upsert_active_entity)",
                    "items": {
                        "type": "object",
                        "properties": {
                            "tool": {"type": "string"},
                            "args": {"type": "object"},
                        },
                        "required": ["tool"],
                    },
                    "default": [],
                },
            },
        },
    }


# ─── Layer 3a: token 黑名单 ──────────────────────────────────────

def validator_token_blacklist(
    proposal: dict, snapshot: dict, script_overrides: dict | None = None
) -> tuple[bool, str]:
    """检查 proposal 文本是否含跨 phase 的不允许 token。

    从 modules/_script_overrides/<key>.json 的 phase_inference rules 推断:
    如果 proposal.summary 含其他 phase 的 needles,reject。

    返回 (passed, reason)。
    """
    summary = (proposal.get('summary') or '').strip()
    if not summary:
        return True, ""  # empty summary 算 noop

    if not script_overrides:
        return True, ""  # 无 overrides 通过 (通用底座默认允许)

    current_phase = snapshot.get('current_phase') or ''
    if not current_phase:
        return True, ""  # 无 phase 信息无法验证

    rules = (script_overrides.get('phase_inference') or {}).get('rules') or []
    # 找当前 phase 之外的 needles → 黑名单
    blacklist: list[str] = []
    for rule in rules:
        if rule.get('phase') and rule['phase'] != current_phase:
            blacklist.extend(rule.get('or_text_needles') or [])

    for token in blacklist:
        if token in summary:
            return False, f"含跨 phase token '{token}' (当前 {current_phase})"
    return True, ""


# ─── Layer 3c: 硬约束 critic (简化版) ─────────────────────────────

def validator_hard_constraints(proposal: dict, snapshot: dict) -> tuple[bool, str]:
    """检查 proposal 是否违反 user locked variables。

    locked_variables 是玩家用 /set 锁定的硬约束,黑天鹅事件不得违反。

    简化版: 检查 summary 是否包含某 locked var key 的反义/否定模式。
    完整版应接 LLM critic,这里留接口。
    """
    locked = snapshot.get('locked_variables') or {}
    if not locked:
        return True, ""

    summary = (proposal.get('summary') or '').strip()
    if not summary:
        return True, ""

    # 简化: 如果 locked var 的 value 在 summary 中被否定 (出现"不"/"没"/"非"近邻)
    for key, value in locked.items():
        if not value:
            continue
        # 粗略匹配: value 后或前 5 字内出现否定词
        for neg in ("不", "没", "非", "未"):
            pat = re.compile(rf"{neg}.{{0,3}}{re.escape(str(value))}|{re.escape(str(value))}.{{0,3}}{neg}")
            if pat.search(summary):
                return False, f"violates locked var: {key}={value}"
    return True, ""


# ─── Layer 3d: timeline 锚点 ─────────────────────────────────────

def validator_timeline_anchor(proposal: dict, snapshot: dict) -> tuple[bool, str]:
    """检查 proposal 涉及的 NPC 是否都在 active_npcs 列表里 (即当前 phase 在场)。

    proposal.involved_npcs 必须 ⊆ snapshot.active_npcs.ids。
    """
    proposed_npcs = set(proposal.get('involved_npcs') or [])
    if not proposed_npcs:
        return True, ""

    active_ids = {n['id'] for n in snapshot.get('active_npcs', []) if n.get('id')}

    invalid = proposed_npcs - active_ids
    if invalid:
        return False, f"涉及未在场 NPC: {invalid}"
    return True, ""


# ─── Layer 3b/3e: 留接口 (TODO) ──────────────────────────────────

def validator_npc_presence(
    proposal: dict, snapshot: dict, script_id: int | None
) -> tuple[bool, str]:
    """TODO: 接 script_character_cards.available_in_phases 字段。

    当前 short-circuit 通过。等剧本元数据扩展后实现。
    """
    return True, ""  # TODO


def validator_independent_critic(proposal: dict, snapshot: dict) -> tuple[bool, str]:
    """TODO: 接 independent LLM critic (二次 LLM 评分一致性)。

    当前 short-circuit 通过。需要 prompt tuning + LLM 调用。
    """
    return True, ""  # TODO


# ─── 全套 validator 跑 ────────────────────────────────────────────

def run_validators(
    proposal: dict, snapshot: dict,
    script_id: int | None, script_overrides: dict | None
) -> list[tuple[str, bool, str]]:
    """跑所有 validator,返回 [(name, passed, reason), ...]"""
    return [
        ("3a_token_blacklist", *validator_token_blacklist(proposal, snapshot, script_overrides)),
        ("3b_npc_presence", *validator_npc_presence(proposal, snapshot, script_id)),
        ("3c_hard_constraints", *validator_hard_constraints(proposal, snapshot)),
        ("3d_timeline_anchor", *validator_timeline_anchor(proposal, snapshot)),
        ("3e_independent_critic", *validator_independent_critic(proposal, snapshot)),
    ]


# ─── Layer 5: dispatcher 落地 ───────────────────────────────────

def dispatch_event(
    proposal: dict, state: Any,
    user_id: int, save_id: int, script_id: int | None,
) -> list[dict]:
    """把 proposal.tools_to_call 通过 dispatcher 落地。

    origin="autonomous_agent", trace_id="swan-<token>"。
    """
    tools_to_call = proposal.get('tools_to_call') or []
    if not tools_to_call:
        return []

    from tools_dsl.command_dispatcher import (
        ToolCallEnvelope,
        ToolDispatcher,
        get_registry,
    )

    trace_id = f"swan-{secrets.token_urlsafe(6)}"
    _sp = cast(Callable[[Any], Any], lambda env, _s=state: _s)
    dispatcher = ToolDispatcher(
        registry=get_registry(),
        state_provider=_sp,
    )
    results = []
    for call in tools_to_call:
        tool_name = call.get('tool') or ""
        args = call.get('args') or {}
        if not tool_name:
            continue
        env = ToolCallEnvelope(
            user_id=user_id,
            save_id=save_id,
            script_id=script_id,
            tool=tool_name,
            args=args,
            origin="autonomous_agent",
            trace_id=trace_id,
        )
        res = dispatcher.dispatch_sync(env)
        results.append({
            "tool": tool_name,
            "ok": res.ok,
            "result": res.result if res.ok else res.error,
        })
    return results


# ─── 入口: maybe_trigger ─────────────────────────────────────────

def maybe_trigger(
    state: Any,
    *,
    user_id: int,
    save_id: int,
    script_id: int | None = None,
    max_retries: int = 2,
    llm_caller: Any | None = None,
) -> dict:
    """post-GM hook 调用入口。

    返回结果 dict:
      - triggered: bool (是否成功产生 + dispatch 了 black swan event)
      - proposal: dict (最终被接受的 proposal,或最后一次 reject 的 proposal)
      - validator_results: list (Layer 3 检查结果)
      - dispatch_results: list (Layer 5 落地结果)
      - retries: int (实际重试次数)
      - reason: str (跳过/失败原因)
    """
    snapshot = reality_snapshot(state, script_id)

    # 无 LLM 调用器 → 跳过 (允许测试时禁用)
    if llm_caller is None:
        return {
            "triggered": False,
            "reason": "no llm_caller provided (test mode or disabled)",
            "snapshot": snapshot,
        }

    # 加载 script overrides (Layer 3a 需要)
    script_overrides = None
    if snapshot.get('current_phase'):
        try:
            from state.core import _load_script_overrides
            all_overrides = _load_script_overrides()
            # 找含当前 phase 的 override
            for key, ov in all_overrides.items():
                rules = (ov.get('phase_inference') or {}).get('rules') or []
                if any(r.get('phase') == snapshot['current_phase'] for r in rules):
                    script_overrides = ov
                    break
        except Exception:
            pass

    schema = proposal_tool_schema(snapshot)

    last_proposal: dict | None = None
    last_validator_results: list = []

    for attempt in range(max_retries + 1):
        # Layer 2: LLM call with strong schema
        try:
            proposal = llm_caller(
                snapshot, schema,
                prev_failure=last_validator_results if attempt > 0 else None
            )
        except Exception as e:
            return {
                "triggered": False,
                "reason": f"llm_caller failed: {e}",
                "snapshot": snapshot,
                "retries": attempt,
            }

        if not proposal or proposal.get('event_kind') == 'no_op':
            return {
                "triggered": False,
                "reason": "llm chose no_op",
                "proposal": proposal,
                "snapshot": snapshot,
                "retries": attempt,
            }

        last_proposal = proposal
        # Layer 3: validators
        last_validator_results = run_validators(proposal, snapshot, script_id, script_overrides)
        all_passed = all(v[1] for v in last_validator_results)

        if all_passed:
            # Layer 5: dispatch
            dispatch_results = dispatch_event(
                proposal, state,
                user_id=user_id, save_id=save_id, script_id=script_id,
            )
            return {
                "triggered": True,
                "proposal": proposal,
                "validator_results": last_validator_results,
                "dispatch_results": dispatch_results,
                "retries": attempt,
                "snapshot": snapshot,
            }
        # Layer 4: retry feedback — loop continues

    # max_retries 用完
    return {
        "triggered": False,
        "reason": f"validators rejected after {max_retries + 1} attempts",
        "proposal": last_proposal,
        "validator_results": last_validator_results,
        "retries": max_retries + 1,
        "snapshot": snapshot,
    }
